Sources/Components/CarouselView/CarouselView.swift

```diff
@@ -0,0 +1,192 @@
+
+    func scrollToItem(at indexPath: IndexPath, at scrollPosition: UICollectionView.ScrollPosition, animated: Bool) {
+        collectionView.scrollToItem(at: IndexPath(item: indexPath.item + 1, section: 0), at: scrollPosition, animated: animated)
+    }
+}
+
+extension CarouselView: UICollectionViewDataSource {
+    public func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
+        guard let dataSource = dataSource else {
+            return 0
+        }
+
+        return dataSource.numberOfItems(in: self) + 2
+    }
+
+    public func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
+        guard let dataSource = dataSource else {
+            preconditionFailure("Data source is not available")
+        }
+
+        return dataSource.carouselView(self, cellForItemAt: self.indexPath(forItem: indexPath.item))
+    }
+}
+
+extension CarouselView: UICollectionViewDelegateFlowLayout {
+    public func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, sizeForItemAt indexPath: IndexPath) -> CGSize {
+        return bounds.size
+    }
+
+    public func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
+        delegate?.carouselView(self, didSelectItemAt: self.indexPath(forItem: indexPath.item))
+    }
+
+    public func scrollViewDidEndDecelerating(_ scrollView: UIScrollView) {
+        delegate?.carouselViewDidEndDecelerating(self)
+    }
+
+    public func scrollViewWillEndDragging(_ scrollView: UIScrollView, withVelocity velocity: CGPoint, targetContentOffset: UnsafeMutablePointer<CGPoint>) {
+        let numberOfItems = collectionView.numberOfItems(inSection: 0)
+
+        if scrollView.isPagingEnabled {
+            let targetPage = Int(targetContentOffset.pointee.x / scrollView.bounds.width)
+            let deltaOffset = CGFloat(numberOfItems - 2) * scrollView.bounds.width
+
+            if targetPage == 0 {
+                targetContentOffset.pointee = CGPoint(
+                    x: deltaOffset,
+                    y: targetContentOffset.pointee.y
+                )
+
+                scrollView.contentOffset = CGPoint(
+                    x: scrollView.contentOffset.x + deltaOffset,
+                    y: scrollView.contentOffset.y
+                )
+
+            } else if targetPage == numberOfItems - 1 {
+                targetContentOffset.pointee = CGPoint(
+                    x: scrollView.bounds.width,
+                    y: targetContentOffset.pointee.y
+                )
+
+                scrollView.contentOffset = CGPoint(
+                    x: scrollView.contentOffset.x - deltaOffset,
+                    y: scrollView.contentOffset.y
+                )
+            }
+        }
+    }
+}
+
+// MARK: - Private Functions
+
+private extension CarouselView {
+    func indexPath(forItem item: Int) -> IndexPath {
```

Comment:
What would happen if the number of items is 0?

Reply:
Something would be definitely break 😅 This carousel view doesn't make sense unless there are more than one data entry so I can return the same item is number of items is not sufficient.